include/inexor/vulkan-renderer/wrapper/descriptor_builder.hpp

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <vulkan/vulkan_core.h>
+
+#include <cassert>
+#include <string>
+#include <vector>
+
+namespace inexor::vulkan_renderer::wrapper {
+
+class Device;
+class ResourceDescriptor;
+
+class DescriptorBuilder {
+    const Device &m_device;
+    const std::uint32_t m_swapchain_image_count;
+
+    std::vector<VkDescriptorSetLayoutBinding> m_layout_bindings;
+    std::vector<VkWriteDescriptorSet> m_write_sets;
+    std::vector<VkDescriptorBufferInfo> m_descriptor_buffer_infos;
+    std::vector<VkDescriptorImageInfo> m_descriptor_image_infos;
+
+public:
+    /// @brief Constructs the descriptor builder.
+    /// @param device The const reference to a device RAII wrapper instance.
+    /// @param swapchain_image_count The number of images in swapchain.
+    DescriptorBuilder(const Device &device, std::uint32_t swapchain_image_count);
+
+    DescriptorBuilder(const DescriptorBuilder &) = delete;
+    DescriptorBuilder(DescriptorBuilder &&) = delete;
+    ~DescriptorBuilder() = default;
+
+    DescriptorBuilder &operator=(const DescriptorBuilder &) = delete;
+    DescriptorBuilder &operator=(DescriptorBuilder &&) = delete;
+
+    // TODO: Implement more descriptor types than just uniform buffers and combined image samplers.
+    // TODO: Support uniform buffer offset in VkDescriptorBufferInfo.
+    // TODO: Offer overloaded methods which expose more fields of the structures.
+
+    /// @brief Adds a uniform buffer to the descriptor container.
+    /// @tparam T The type of the uniform buffer.
+    /// @param uniform_buffer The uniform buffer which contains the data which will be accessed by the shader.
+    /// @param binding The binding index which will be used in the SPIR-V shader.
+    /// @param shader_stage The shader stage the uniform buffer will be used in, most likely the vertex shader.
+    /// @return A const reference to this DescriptorBuilder instance.
+    template <typename T>
+    DescriptorBuilder &add_uniform_buffer(const VkBuffer uniform_buffer, const std::uint32_t binding,
```

Comment:
Please split declaration and definitions.

Reply:
Done.

Reply:
You can
```cpp
    template <typename T>
    DescriptorBuilder &add_uniform_buffer(const VkBuffer uniform_buffer, const std::uint32_t binding,
                                          const VkShaderStageFlagBits shader_stage = VK_SHADER_STAGE_VERTEX_BIT);
// class ends here

template <typename T>
DescriptorBuilder &DescriptorBuilder::add_uniform_buffer(const VkBuffer uniform_buffer, const std::uint32_t binding,
                                          const VkShaderStageFlagBits shader_stage = VK_SHADER_STAGE_VERTEX_BIT) {
    // code goes here
}
```

Reply:
Done.

Reply:
Done.